config/configparser/parser.go

```diff
@@ -17,96 +17,118 @@ package configparser
 import (
 	"fmt"
 	"io"
+	"io/ioutil"
 	"reflect"
-	"strings"
 
+	"github.com/knadh/koanf"
+	"github.com/knadh/koanf/parsers/yaml"
+	"github.com/knadh/koanf/providers/confmap"
+	"github.com/knadh/koanf/providers/file"
+	"github.com/knadh/koanf/providers/rawbytes"
+	"github.com/mitchellh/mapstructure"
 	"github.com/spf13/cast"
-	"github.com/spf13/viper"
 )
 
 const (
-	// KeyDelimiter is used as the default key delimiter in the default viper instance.
+	// KeyDelimiter is used as the default key delimiter in the default koanf instance.
 	KeyDelimiter = "::"
 )
 
 // NewParser creates a new empty Parser instance.
 func NewParser() *Parser {
-	return &Parser{
-		v: viper.NewWithOptions(viper.KeyDelimiter(KeyDelimiter)),
-	}
+	return &Parser{k: koanf.New(KeyDelimiter)}
 }
 
 // NewParserFromFile creates a new Parser by reading the given file.
 func NewParserFromFile(fileName string) (*Parser, error) {
 	// Read yaml config from file.
 	p := NewParser()
-	p.v.SetConfigFile(fileName)
-	if err := p.v.ReadInConfig(); err != nil {
+	if err := p.k.Load(file.Provider(fileName), yaml.Parser()); err != nil {
 		return nil, fmt.Errorf("unable to read the file %v: %w", fileName, err)
 	}
 	return p, nil
 }
 
 // NewParserFromBuffer creates a new Parser by reading the given yaml buffer.
 func NewParserFromBuffer(buf io.Reader) (*Parser, error) {
+	content, err := ioutil.ReadAll(buf)
+	if err != nil {
+		return nil, err
+	}
+
 	p := NewParser()
-	p.v.SetConfigType("yaml")
-	if err := p.v.ReadConfig(buf); err != nil {
+	if err := p.k.Load(rawbytes.Provider(content), yaml.Parser()); err != nil {
 		return nil, err
 	}
+
 	return p, nil
 }
 
 // NewParserFromStringMap creates a parser from a map[string]interface{}.
 func NewParserFromStringMap(data map[string]interface{}) *Parser {
 	p := NewParser()
-	// Cannot return error because the viper instance is empty.
-	_ = p.v.MergeConfigMap(data)
+	// Cannot return error because the koanf instance is empty.
+	_ = p.k.Load(confmap.Provider(data, KeyDelimiter), nil)
```

Comment:
Does this comment still apply?

Reply:
I'm not clear tbh and thought it was more of a reference to the signature not also returning an error.

Reply:
Maybe we should return an error and be done with this. Can be done in a followup.